src/constitutive/TACSBladeStiffenedShellConstitutive.cpp

```diff
@@ -2853,6 +3066,552 @@ bool TACSBladeStiffenedShellConstitutive::testBucklingEnvelopeSens(
+
+  const TacsScalar columnBucklingLoad =
+      this->computeStiffenerColumnBucklingLoad();
+  const TacsScalar stiffenerAxialLoad = -stiffenerStress[0];
+  const TacsScalar fCrit = this->computeStiffenerColumnBucklingLoad();
+  memset(sens, 0, numStress * sizeof(TacsScalar));
+
+  for (int ii = 0; ii < numStress; ii++) {
+    sens[ii] = -C[ii] / fCrit;
+  }
+  return stiffenerAxialLoad / fCrit;
+}
+
+void TACSBladeStiffenedShellConstitutive::addStiffenerColumnBucklingDVSens(
+    const TacsScalar scale, const TacsScalar shellStrain[],
+    const TacsScalar stiffenerStrain[], const TacsScalar stiffenerAxialLoad,
+    const TacsScalar fCrit, TacsScalar dfdx[]) {
+  // F = stiffenerAxialLoad / fCrit
+  // Using the Quotient rule:
+  // dF/dx =  1/fCrit * dS11/dx - stiffenerAxialLoad/fCrit^2 * dfCrit/dx
+
+  // First we add the 1/fCrit * dS11/dx term
+  // We need to consider both the direct dependence of the stiffener stress on
+  // the DVs, and also the dependence of the stiffener stress on the stiffener
+  // centroid strains, which depend on the DVS:
+  // psi^T dS/dx = psi^T pS/px + psi^T pS/pe * dstiffenerStrain/dx
+  // = psi^T pS/px + C * psi * d/dx(Te) * shellStrain
+  TacsScalar psi[TACSBeamConstitutive::NUM_STRESSES];
+  memset(psi, 0, TACSBeamConstitutive::NUM_STRESSES * sizeof(TacsScalar));
+  psi[0] = 1.0;
+  this->addStiffenerStressDVSens(-scale / fCrit, stiffenerStrain, psi,
+                                 &dfdx[this->stiffenerDVStartNum]);
+
+  TacsScalar psiStress[TACSBeamConstitutive::NUM_STRESSES];
+  this->computeStiffenerStress(psi, psiStress);
+  this->addStrainTransformProductDVsens(psiStress, shellStrain, -scale / fCrit,
+                                        dfdx);
+
+  // Now we add the - stiffenerAxialLoad/fCrit^2 * dfCrit/dx term
+  this->addStiffenerColumnBucklingLoadDVSens(
+      -scale * stiffenerAxialLoad / (fCrit * fCrit), dfdx);
+}
+
+void TACSBladeStiffenedShellConstitutive::computeStiffenerCripplingValues(
+    const TacsScalar stiffenerStrain[], TacsScalar plyFailValues[]) {
+  // We use the semi-empirical formula for one-edge-free crippling described
+  // in the DoD composite materials handbook Volume 3F (a.k.a MIL-HDBK-17-3F)
+  // (available at
+  // http://everyspec.com/MIL-HDBK/MIL-HDBK-0001-0099/MIL_HDBK_17_3F_216/)
+  //
+  // The formula is: F_crippling / F_ult = 1.63*(b/t)^-0.717
+  // Where:
+  // F_crippling is the crippling load
+  // F_ult is the ultimate compressive load
+  // b is the flange width
+  // t is the flange thickness
+  //
+  // Here we compute the ultimate compressive load using the same criteria
+  // used to compute the stiffener material failure value but using on the
+  // axial strain component. This gives: SR_Crippling = SR_comp/1.63 *
+  // (b/t)^0.717 Where SR_Crippling is the crippling strength ratio SR_comp is
+  // the strength ratio computed using only the compressive component of the
+  // strain
+
+  const int numPlies = this->numStiffenerPlies;
```

Comment:
Have you tested the stiffener crippling critical loads here to ensure that they give reasonable limits on stiffener aspect ratio? Could use a simple flat panel example and axial loads just to check.